src/server/api/consume-wifi-data.ts

```diff
@@ -0,0 +1,109 @@
+import { Kafka } from "kafkajs";
+import { SchemaRegistry } from "@kafkajs/confluent-schema-registry";
+import { serverSupabaseClient } from "#supabase/server";
+import type { SupabaseClient } from '@supabase/supabase-js';
+
+const { internalSecret, kafkaConfig, schemaRegistry } = useRuntimeConfig();
+
+console.time("Initialize");
+const kafka = new Kafka({
+  clientId: "tudelft-spacefinder",
+  ssl: true,
+  sasl: {
+    mechanism: "plain",
+    username: kafkaConfig.username,
+    password: kafkaConfig.password,
+  },
+  brokers: [kafkaConfig.url],
+});
+
+const registry = new SchemaRegistry({
+  host: schemaRegistry.url,
+  auth: {
+    username: schemaRegistry.username,
+    password: schemaRegistry.password,
+  },
+});
+
+const consumer = kafka.consumer({ groupId: kafkaConfig.groupId });
+console.timeEnd("Initialize");
+
+let seeked = false;
+
+export default defineEventHandler(async (event) => {
+  if (getQuery(event)?.secret !== internalSecret) {
+    event.node.res.statusCode = 401;
+    event.node.res.end();
+    return;
+  }
+
+  const client = serverSupabaseClient(event);
+
+  consumeLastBatch({ client });
+
+  // Close consumer connection within function time limit
+  await new Promise((resolve) => setTimeout(async () => {
+    console.time("Consumer disconnect");
+    await consumer.disconnect();
+    console.timeEnd("Consumer disconnect");
+    event.node.res.statusCode = 202;
+    event.node.res.end();
+    resolve(null);
+  }, 20000));
+});
+
+async function consumeLastBatch({ client }: { client: SupabaseClient }) {
+  console.time("Consumer setup");
+  await consumer.connect();
+  await consumer.subscribe({ topic: kafkaConfig.topic });
+  console.timeEnd("Consumer setup");
+
+  consumer.run({
+    eachBatch: async ({ batch }) => {
+      if (!seeked) {
+        seeked = true;
+        consumer.seek({
+          topic: kafkaConfig.topic,
+          offset: batch.highWatermark,
+          partition: batch.partition,
+        });
+      }
+
+      console.time("Parse messages");
+      const parsedMessages = await Promise.all(
+        batch.messages
+          .filter((message) => message.value)
+          .map((message) =>
+            registry
+              .decode(message.value)
+              .then((decodedValue) => ({
+                "updated_at": new Date(Number(message.timestamp)).toISOString(),
+                "access_point_name": decodedValue.name,
+                "device_count": decodedValue.clientCount,
+                "building_number": decodedValue.name.split("-").at(1) || 9001,
```

Comment:
With this exact comment I meant the special case where the building number cannot be deducted from `name` and should come from `mapLocation` instead (you found that case yourself). As for 9001, it could be replaced by null I guess, but that would at least require us to rewrite the triggers a little bit. Since it won't ever fall back to null in real life provided the data we have now, I suggest we do it in a separate PR. Maybe together will logging this case, idk

Reply:
👍 ah gotcha. I've added the maplocation fallback.
Let's discuss with Jasper tomorrow at the daily if we want to log this or use null.